### PR TITLE
Add semtype tests to testSuite

### LINKED ISSUE #499: Improve semtypes testing
Need to improve both tests and test framework.

### PR DESCRIPTION
**Pull request by ushirask**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/showTypes.bal [hunk 1]
```diff
@@ -39,4 +39,46 @@ function subtypeRels(front:SourcePart[] sources) returns string[]|err:Any|io:Err
         order by s
         select s;
     return rels;
-}
\ No newline at end of file
+}
+
+function testSubtypes(front:SourcePart[] sources, string[] expected) returns boolean|err:Any|io:Error {
+
```
Comment (manuranga) [inline compiler/showTypes.bal]: this repo don't have the convention of leaving an empty line at the top.

File: compiler/showTypes.bal [hunk 2]
```diff
@@ -39,4 +39,45 @@ function subtypeRels(front:SourcePart[] sources) returns string[]|err:Any|io:Err
...
-}
\ No newline at end of file
+}
+
+function testSubtypes(front:SourcePart[] sources, string[] expected) returns boolean|err:Any|io:Error {
+    var [env, m] = check front:typesFromString(sources);
+    var tc = t:typeContext(env);
+    foreach var item in expected {
+        int? i1 = item.indexOf(" < ");
+        if i1 is int {
```
Comment (manuranga) [inline compiler/showTypes.bal]: Lot of logic in these three `if`s seems to be repeating, please deduplicate via functions.

File: compiler/modules/front.syntax/parseTypeTest.bal [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        }  
+        
+    }
+    return parseError(tok);
+}
+
+public type SubtypeTest record {
+    SubtypeTestOp op;
+    string typename1;
+    string typename2;
```
Comment (manuranga) [inline compiler/modules/front.syntax/parseTypeTest.bal]: lets use `left` and `right`.

File: compiler/modules/front.syntax/parseTypeTest.bal [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import wso2/nballerina.err;
+
+public function parseTypeTest(string str) returns SubtypeTest|err:Syntax {
+    SourceFile file = createSourceFile([str], { filename: "<internal>" });
+    Tokenizer tok = new(file);
+    check tok.advance();
+    Token? t = tok.current();
+    // TODO move to different parse functions
+    if t is [IDENTIFIER, string] {
```
Comment (manuranga) [inline compiler/modules/front.syntax/parseTypeTest.bal]: please use  tok.expectIdentifier();

**Comment by manuranga** [comment]
I didn't expect you to use the `Tokenizer` to parse the line, but now that you have used it, I see no problem with it.

File: compiler/showTypes.bal [hunk 5]
```diff
@@ -39,4 +41,38 @@ function subtypeRels(front:SourcePart[] sources) returns string[]|err:Any|io:Err
...
\ No newline at end of file
+}
+
+function testSubtypes(front:SourcePart[] sources, string[] expected) returns boolean|err:Syntax|err:Any|io:Error {
+    var [env, m] = check front:typesFromString(sources);
+    var tc = t:typeContext(env);
+    foreach var item in expected {
+        s:SubtypeTest test = check s:parseTypeTest(item);
+        var [t1, t2] = extractSemtypes(test.typename1, test.typename2, m);        
+        if test.op is "<" {
```
Comment (manuranga) [inline compiler/showTypes.bal]: Use `match`. I don't think you'll need continue then.

File: compiler/modules/front.syntax/parseTypeTest.bal [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    t = tok.current();
+    if t is "<"|"=" {
+        SubtypeTestOp op = t;
+        check tok.advance();
+        t = tok.current();
+        if t is ">" {
+            op = "<>";
+            check tok.advance();
+            t = tok.current();
+        }
```
Comment (manuranga) [inline compiler/modules/front.syntax/parseTypeTest.bal]:
```suggestion
        check tok.advance();
        SubtypeTestOp op;
        if tok.current() is ">" {
            check tok.advance();
            op = "<>";
        }
        else {
            op = t;
        }
```
(Please think about why I am proposing this alternative)

File: compiler/showTypes.bal [hunk 7]
```diff
@@ -39,4 +41,37 @@ function subtypeRels(front:SourcePart[] sources) returns string[]|err:Any|io:Err
...
+    var [env, m] = check front:typesFromString(sources);
+    var tc = t:typeContext(env);
+    foreach var item in expected {
+        s:SubtypeTest test = check s:parseTypeTest(item);
+        var [t1, t2] = extractSemtypes(test.left, test.right, m);
+        match test.op { 
+            "<" => {
+                if !t:isSubtype(tc, t1, t2) || t:isSubtype(tc, t2, t1) {
+                    return false;
+                }
```
Comment (manuranga) [inline compiler/showTypes.bal]: This should be an `test:assertFalse` and a `test:assertTrue`.

Comment (ushirask) [reply compiler/showTypes.bal]: is there a disadvantage/limitation in carrying over a boolean to testSuite and asserting there?

Comment (manuranga) [reply compiler/showTypes.bal]: When something fails, we need to know which exact check that failed. If we return a boolean form here and assert elsewhere, we'll just get to know something went wrong, but not the exact details.

In fact it's a good idea to write couple of failing tests and see if the error messages are meaningful.

Comment (manuranga) [reply compiler/showTypes.bal]: Did  you test a filing case with multiple assertions? do we get a good error message?

Comment (ushirask) [reply compiler/showTypes.bal]: Yes, I believe we'll be able to adequately debug with the error messages

File: compiler/showTypes.bal [hunk 8]
```diff
@@ -39,4 +42,28 @@ function subtypeRels(front:SourcePart[] sources) returns string[]|err:Any|io:Err
...
+    foreach var item in expected {
+        s:SubtypeTest test = check s:parseTypeTest(item);
+        t:SemType t1 = m.entries().get(test.left)[1];
+        t:SemType t2 = m.entries().get(test.right)[1];
+        match test.op { 
+            "<" => {
+                    test:assertTrue(t:isSubtype(tc, t1, t2), test.left + " is not a subtype of " + test.right);
+                    test:assertFalse(t:isSubtype(tc, t2, t1), test.right + " is a subtype of " + test.left);
+                }
+                "<>" => {
```
Comment (jclark) [inline compiler/showTypes.bal]: Formatted wrong.

File: compiler/showTypes.bal [hunk 9]
```diff
@@ -39,4 +42,28 @@ function subtypeRels(front:SourcePart[] sources) returns string[]|err:Any|io:Err
         order by s
         select s;
     return rels;
-}
\ No newline at end of file
+}
+
+function testSubtypes(front:SourcePart[] sources, string[] expected) returns err:Syntax|err:Any|io:Error? {
```
Comment (jclark) [inline compiler/showTypes.bal]: This should be in the tests/ dir.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/showTypes.bal (hunk 9 above)
Comment (ushirask) [reply compiler/showTypes.bal]: moved to testSuite

File: compiler/tests/testSuite.bal [hunk 10]
```diff
@@ -27,6 +30,14 @@ function testCompileVPO(string path, string kind) returns io:Error? {
     }
 }
 
+@test:Config {
+    dataProvider: listSourcesT
+}
+function testSemTypeT(string path, string kind) returns error? {
```
Comment (manuranga) [inline compiler/tests/testSuite.bal]: Prefer to call it just `testSemType`.

**Comment by ushirask** [comment]
Closing since this is included in #534

**Pull request closed without merging**